File: Sep/Day17/design-a-food-rating-system.js
class FoodRatings {
    constructor(foods, cuisines, ratings) {
        this.foodMap = new Map(); // food -> [rating, cuisine]
        this.cuisineMap = new Map(); // cuisine -> heap (array used as priority queue)

        for (let i = 0; i < foods.length; i++) {
            let food = foods[i], cuisine = cuisines[i], rating = ratings[i];
            this.foodMap.set(food, [rating, cuisine]);

            if (!this.cuisineMap.has(cuisine)) this.cuisineMap.set(cuisine, []);
            this._push(this.cuisineMap.get(cuisine), [-rating, food]); // store as minHeap by neg rating
        }
    }

    changeRating(food, newRating) {
        let [oldRating, cuisine] = this.foodMap.get(food);
        this.foodMap.set(food, [newRating, cuisine]);
        this._push(this.cuisineMap.get(cuisine), [-newRating, food]); // lazy update
    }

    highestRated(cuisine) {
        let heap = this.cuisineMap.get(cuisine);
        while (heap.length > 0) {
            let [negRating, food] = heap[0];
            let [actualRating] = this.foodMap.get(food);
            if (-negRating === actualRating) return food;
            this._pop(heap); // remove stale
        }
        return "";
    }

    // ---------- MinHeap helper functions ----------
    _push(heap, val) {
        heap.push(val);
        this._siftUp(heap, heap.length - 1);
    }
    _pop(heap) {
        let last = heap.pop();
        if (heap.length === 0) return last;
        let top = heap[0];
        heap[0] = last;
        this._siftDown(heap, 0);
        return top;
    }
    _siftUp(heap, idx) {
        while (idx > 0) {
            let parent = Math.floor((idx - 1) / 2);
            if (this._cmp(heap[idx], heap[parent]) >= 0) break;
            [heap[idx], heap[parent]] = [heap[parent], heap[idx]];
            idx = parent;
        }
    }
    _siftDown(heap, idx) {
        let n = heap.length;
        while (true) {
            let left = idx * 2 + 1, right = idx * 2 + 2, smallest = idx;
            if (left < n && this._cmp(heap[left], heap[smallest]) < 0) smallest = left;
            if (right < n && this._cmp(heap[right], heap[smallest]) < 0) smallest = right;
            if (smallest === idx) break;
            [heap[idx], heap[smallest]] = [heap[smallest], heap[idx]];
            idx = smallest;
        }
    }
    _cmp(a, b) {
        if (a[0] !== b[0]) return a[0] - b[0]; // compare neg ratings
        return a[1].localeCompare(b[1]); // lexicographic
    }
}

File: Sep/Day17/design-a-food-rating-system.java
import java.util.*;

class FoodRatings {
    static class Pair {
        int rating;
        String cuisine;
        Pair(int r, String c) { rating = r; cuisine = c; }
    }

    Map<String, Pair> foodMap; // food -> (rating, cuisine)
    Map<String, PriorityQueue<String>> cuisineMap; // cuisine -> heap of foods

    public FoodRatings(String[] foods, String[] cuisines, int[] ratings) {
        foodMap = new HashMap<>();
        cuisineMap = new HashMap<>();

        for (int i = 0; i < foods.length; i++) {
            String food = foods[i], cuisine = cuisines[i];
            int rating = ratings[i];
            foodMap.put(food, new Pair(rating, cuisine));

            cuisineMap.putIfAbsent(cuisine, new PriorityQueue<>((a, b) -> {
                int r1 = foodMap.get(a).rating, r2 = foodMap.get(b).rating;
                if (r1 != r2) return r2 - r1; // higher rating first
                return a.compareTo(b); // lexicographically smaller first
            }));
            cuisineMap.get(cuisine).offer(food);
        }
    }

    public void changeRating(String food, int newRating) {
        Pair p = foodMap.get(food);
        p.rating = newRating;
        String cuisine = p.cuisine;
        cuisineMap.get(cuisine).offer(food); // lazy insert
    }

    public String highestRated(String cuisine) {
        PriorityQueue<String> heap = cuisineMap.get(cuisine);
        while (!heap.isEmpty()) {
            String topFood = heap.peek();
            Pair p = foodMap.get(topFood);
            boolean valid = true;
            for (String candidate : heap) {
                Pair q = foodMap.get(candidate);
                if ((q.rating > p.rating) || 
                    (q.rating == p.rating && candidate.compareTo(topFood) < 0)) {
                    valid = false;
                    break;
                }
            }
            if (valid) return topFood;
            heap.poll();
        }
        return "";
    }
}
